src/js/components/DataTableColumns/DataTableColumns.js

```diff
@@ -0,0 +1,163 @@
+  // as needed when rendering.
+  const objectOptions = useMemo(
+    () => options && options.length && typeof options[0] === 'object',
+    [options],
+  );
+
+  // 'value' is an array of property names
+  const [value, setValue] = useFormInput({
+    name: formColumnsKey,
+    initialValue: optionsToValue(options?.slice(0, 3)),
+  });
+
+  // When the user searches, updated the filtered options based on the
+  // search string.
+  const onSearch = useCallback(
+    (nextSearch) => {
+      let nextFilteredOptions = options;
+      if (nextSearch) {
+        const lowerSearch = nextSearch.toLowerCase();
+        nextFilteredOptions = options.filter(
+          (o) =>
+            (!objectOptions && o.toLowerCase().includes(lowerSearch)) ||
+            (objectOptions &&
+              (o.property.toLowerCase().includes(lowerSearch) ||
+                o.label.toLowerCase().includes(lowerSearch))),
+        );
+      }
+      setSearch(nextSearch);
+      setFilteredOptions(nextFilteredOptions);
+    },
+    [objectOptions, options],
+  );
+
+  return (
+    <Box>
+      <Tabs {...tabsProps[drop ? 'drop' : 'noDrop']} {...rest}>
+        <Tab
+          id={`${dataId}--select-columns-tab`}
+          title={format({
+            id: 'dataTableColumns.select',
+            messages: messages?.dataTableColumns,
+          })}
+        >
+          <Box pad={{ top: 'small' }}>
+            <TextInput
+              type="search"
+              icon={<Search />}
+              placeholder="Search"
+              value={search}
+              onChange={(event) => onSearch(event.target.value)}
+            />
+            <CheckBoxGroup
+              id={`${dataId}--select-columns`}
+              name={formColumnsKey}
+              aria-labelledby={`${dataId}--select-columns-tab`}
+              options={filteredOptions}
+              valueKey={(objectOptions && 'property') || undefined}
+              labelKey={(objectOptions && 'label') || undefined}
+              value={value}
+              onChange={({ value: nextValue }) => setValue(nextValue)}
+            />
+          </Box>
+        </Tab>
+        <Tab
+          id={`${dataId}--order-columns-tab`}
+          title={format({
+            id: 'dataTableColumns.order',
+            messages: messages?.dataTableColumns,
+          })}
+        >
+          <List
+            id={`${dataId}--order-columns`}
+            aria-labelledby={`${dataId}--order-columns-tab`}
+            // List wants objects if possible to be able to use 'label'
+            data={value.map(
+              (v) =>
+                (objectOptions && options.find((o) => o.property === v)) || v,
+            )}
+            onOrder={(nextData) => setValue(optionsToValue(nextData))}
+            pad="none"
+            primaryKey={(objectOptions && 'label') || undefined}
+          />
+        </Tab>
+      </Tabs>
+    </Box>
+  );
+};
+
+export const DataTableColumns = ({ drop, options, ...rest }) => {
```

Comment:
How could you persist the changes in visibility or order? Give access to onClose? Have an onChange?

Reply:
That functionality is coming in a separate pull request, tied to "named views".